chokidar on demand

at0g opened the pull request:
- make chokidar an optional dependency
- only require chokidar when `{ watch: true }`
- if chokidar is not installed and `{ watch: true }`, notify the user with a console error (but don't fail)

Fixes #600

jgerigmeyer commented:
:+1:

carljm commented on src/node-loaders.js:
```diff
@@ -34,20 +33,26 @@ var FileSystemLoader = Loader.extend({
+                    console.log('Watcher error: ' + error);
+                });
+            }
+            catch (err) {
```
This seems like a really broad catch, which could easily catch some other (unexpected) error, and proceed to give a wrong and misleading message on the console.

Where exactly will an error be raised if chokidar isn't installed? On the `require('chokidar')` line, right? So why not just wrap that line only?

carljm commented:
This could use a note in the documentation for `watch: true`? Other than that and the one code comment I just made, this looks merge-ready to me.

FWIW, personally I see this as the first step on a path to eventually deprecating and removing `watch: true` entirely. I think it's out of scope for a templating language to have its own watch system; that should be left in the realm of gulp et al.

carljm commented:
Even with the outpouring of positive response to #610, I'm still planning to merge this -- it'll help for future releases in the 2.x series, since it's backward-compatible, whereas I can't remove `watch: true` entirely until 3.0.

atian25 commented:
+1 for this.  when will this merge && publish ?

any plan and roadmap for 3.x?

carljm commented:
@atian25 if you file a PR based on the work here that addresses my comments above, I will merge it!  Otherwise I'll get to it when I can, but I already spent quite a few hours on nunjucks this week, so that might not be right away.

carljm commented:
Re 3.x roadmap, I'm beginning to develop one. So far it involves more removals than additions :) mostly it depends on what people are motivated enough to file PRs for, I know I won't have time to personally fix nearly all the open issues.

atian25 commented:
@at0g any plan at this?

at0g commented:
Hello, sorry about the slow reply, been a busy time of year.

I've updated the PR with your feedback:
- Narrow the catch block @jgerigmeyer 
- Added a note to the docs to install chokidar when setting `watch: true` @carljm 

Let me know if there's anything else, otherwise happy new year!

ricordisamoa commented:
If chokidar is not installed and `{ watch: true }`, why not fail fast?

carljm commented:
@at0g This looks great! Unfortunately the linter is complaining; I think it doesn't like the fact that you have the declaration of `var chokidar` inside that `try` block, while the usage of it is outside. Can you just stick a `var chokidar` up one level instead?

@ricordisamoa not sure I understand your comment - this PR does fail as fast as is possible if chokidar is not installed.

ricordisamoa commented:
@carljm Oh sorry, I meant throwing an error instead of logging.

carljm commented:
@ricordisamoa Oh! I actually missed that, thanks for pointing it out. I don't feel strongly, but yeah I think we may as well just throw in that case. There's no reason to proceed with an invalid configuration, the user should just fix it one way or another (install chokidar, or don't set `watch: true`).

TrySound commented:
Any news?

atian25 commented:
@carljm @at0g  any news about this?
